cpp/oneapi/dal/backend/primitives/stat/cov_dpc.cpp

```diff
@@ -0,0 +1,139 @@
+
+#include <CL/sycl/ONEAPI/experimental/builtins.hpp>
+
+namespace oneapi::dal::backend::primitives {
+
+template <typename Float>
+sycl::event correlation(sycl::queue& queue,
+                        const table& data,
+                        const ndview<Float, 1>& sums,
+                        ndview<Float, 2>& corr,
+                        ndview<Float, 1>& means,
+                        ndview<Float, 1>& vars,
+                        ndview<Float, 1>& tmp,
+                        const event_vector& deps) {
+    ONEDAL_ASSERT(data.has_data());
+    ONEDAL_ASSERT(sums.has_data());
+    ONEDAL_ASSERT(corr.has_mutable_data());
+    ONEDAL_ASSERT(means.has_mutable_data());
+    ONEDAL_ASSERT(vars.has_mutable_data());
+    ONEDAL_ASSERT(tmp.has_mutable_data());
+    ONEDAL_ASSERT(corr.get_dimension(0) == corr.get_dimension(1),
+                  "Correlation matrix must be square");
+    ONEDAL_ASSERT(corr.get_dimension(0) == data.get_column_count(),
+                  "Dimensions of correlation matrix must match column count of input data");
+    ONEDAL_ASSERT(sums.get_dimension(0) == data.get_column_count(),
+                  "Element count of sums must match column count of input data");
+    ONEDAL_ASSERT(vars.get_dimension(0) == data.get_column_count(),
+                  "Element count of vars must match column count of input data");
+    ONEDAL_ASSERT(means.get_dimension(0) == data.get_column_count(),
+                  "Element count of means must match column count of input data");
+    ONEDAL_ASSERT(tmp.get_dimension(0) == data.get_column_count(),
+                  "Element count of temporary buffer must match column count of input data");
+
+    // TODO: Determine optimal block size
+    constexpr std::int64_t block_max_row_count = 4096;
+
+    sycl::event gemm_event;
+    array<Float> data_block_flat;
+    const auto data_acc = row_accessor<const Float>{ data };
+
+    for_each_block(data, block_max_row_count, [&](const row_block_info& bi) mutable {
+        data_acc.pull(queue, data_block_flat, bi.get_row_range());
+        const auto x = ndview<Float, 2>::wrap(data_block_flat.get_data(), bi.get_shape());
+
+        // For the first block, call C = X^T x X, beta = 0.0
+        // For the next blocks, call C = X^T x X + C, beta = 1.0
+        const Float beta = Float(bi.get_block_index() != 0);
+        gemm_event = gemm(queue, x.t(), x, corr, Float(1), beta, { gemm_event });
+    });
+
+    const std::int64_t n = data.get_row_count();
+    const std::int64_t p = data.get_column_count();
+    const Float inv_n = Float(1.0 / double(n));
+    const Float inv_n1 = (n > 1) ? Float(1.0 / double(n - 1)) : 1;
+
+    const Float* sums_ptr = sums.get_data();
+    Float* corr_ptr = corr.get_mutable_data();
+    Float* means_ptr = means.get_mutable_data();
+    Float* vars_ptr = vars.get_mutable_data();
+    Float* tmp_ptr = tmp.get_mutable_data();
+
+    auto mean_var_event = queue.submit([&](sycl::handler& cgh) {
```

Comment:
a couple of too lightweight kernels below (if p isn't huge)

Reply:
Yes, I have not idea what to do with them. This is port of that is written in OpenCL.